packages/mui-system/src/Grid/gridGenerator.test.js

```diff
@@ -0,0 +1,397 @@
+      traverseBreakpoints(breakpoints, [1, 2, 3, 4, 5], (appendStyle, value) => {
+        appendStyle(styles, { margin: value });
+      });
+      expect(styles).to.deep.equal({
+        margin: 1,
+        '@media (min-width:600px)': {
+          margin: 2,
+        },
+        '@media (min-width:900px)': {
+          margin: 3,
+        },
+        '@media (min-width:1200px)': {
+          margin: 4,
+        },
+        '@media (min-width:1536px)': {
+          margin: 5,
+        },
+      });
+    });
+
+    it('supports object', () => {
+      const styles = {};
+      traverseBreakpoints(
+        breakpoints,
+        { xs: 1, sm: 2, md: 3, lg: 4, xl: 5 },
+        (appendStyle, value) => {
+          appendStyle(styles, { margin: value });
+        },
+      );
+      expect(styles).to.deep.equal({
+        margin: 1,
+        '@media (min-width:600px)': {
+          margin: 2,
+        },
+        '@media (min-width:900px)': {
+          margin: 3,
+        },
+        '@media (min-width:1200px)': {
+          margin: 4,
+        },
+        '@media (min-width:1536px)': {
+          margin: 5,
+        },
+      });
+    });
+
+    it('works with mixed object', () => {
+      const styles = {};
+      traverseBreakpoints(
+        breakpoints,
+        { a: 2, b: 5, xs: 1, sm: 2, md: 3, foo: () => {}, bar: [] },
+        (appendStyle, value) => {
+          appendStyle(styles, { margin: value });
+        },
+      );
+      expect(styles).to.deep.equal({
+        margin: 1,
+        '@media (min-width:600px)': {
+          margin: 2,
+        },
+        '@media (min-width:900px)': {
+          margin: 3,
+        },
+      });
+    });
+
+    it('does not iterate undefined value', () => {
+      const styles = {};
+      traverseBreakpoints(breakpoints, { xs: 1, sm: undefined, md: 3 }, (appendStyle, value) => {
+        appendStyle(styles, { margin: value });
+      });
+      expect(styles).to.deep.equal({
+        margin: 1,
+        '@media (min-width:900px)': {
+          margin: 3,
+        },
+      });
+    });
+
+    describe('custom breakpoints', () => {
+      const customBreakpoints = createBreakpoints({
+        values: { xxs: 0, xs: 400, sm: 600, md: 768 },
+      });
+
+      it('supports array', () => {
+        const styles = {};
+        traverseBreakpoints(customBreakpoints, [1, 2, 3, 4, 5], (appendStyle, value) => {
+          appendStyle(styles, { margin: value });
+        });
+        expect(styles).to.deep.equal({
+          margin: 1,
+          '@media (min-width:400px)': {
+            margin: 2,
+          },
+          '@media (min-width:600px)': {
+            margin: 3,
+          },
+          '@media (min-width:768px)': {
+            margin: 4,
+          },
+        });
+      });
+
+      it('supports object', () => {
+        const styles = {};
+        traverseBreakpoints(
+          customBreakpoints,
+          { xxs: 1, xs: 2, sm: 3, md: 4, lg: 5 }, // lg is not a part of custom breakpoints
```

Comment:
Let's add test where the breakpoints are not in ascending order?

Reply:
Done.